amethyst_rendy/src/camera.rs

```diff
@@ -671,4 +672,37 @@ mod tests {
         let projected_point = mvp.transform_point(&far);
         assert_abs_diff_eq!(projected_point[2], 0.0);
     }
+
+    #[test]
+    fn screen_ray_3d() {
+        let width = 1280.0;
+        let height = 720.0;
+
+        let aspect = width / height;
+        let fov = std::f32::consts::FRAC_PI_3;
+        let znear = 0.125;
+        let camera = Camera::perspective(aspect, fov, znear);
+
+        let mut camera_transform: Transform = Transform::new(
+            Translation3::new(0.0, 0.0, 3.0),
+            UnitQuaternion::identity(),
+            [1.0, 1.0, 1.0].into(),
+        );
+        camera_transform.copy_local_to_global();
+
+        let cursor_pos = Point2::new(width, height);
+        let screen_diag = Vector2::new(width, height);
+        let ray = camera.screen_ray(cursor_pos, screen_diag, &camera_transform);
+
+        let expected_ray = Ray {
+            /// In the znear plane.
+            origin: Point3::new(0.12830, -0.07216, 2.875),
+            // Corresponds to 45 degree rotation to the right and 30 degree rotation up. This ray
+            // should pass through the top right corner of the screen.
+            direction: Vector3::new(0.66436, -0.37370, -0.64727),
+        };
+        assert_ulps_eq!(ray.origin, expected_ray.origin, max_ulps = 1);
```

Comment:
🤦‍♂️ Okay, that was _completely_ my fault.  I should have checked to see if the asserts worked with `Vector3` ... they do not.

Even with this newest suggestion the expected ray is failing on my mac because it is missing so much precision that the value is quite a ways off (in terms of steps on the floating point scale).  Where did you get the literal values from?

Here are the values on my mac:
- `ray.origin.coords` - `Matrix { data: [0.12830007, -0.0721688, 2.875] }`
- `ray.direction` - `Matrix { data: [0.6643638, -0.37370467, -0.6472754] }`

```suggestion
        assert_ulps_eq!(ray.origin.coords.x, expected_ray.origin.coords.x, max_ulps = 2);
        assert_ulps_eq!(ray.origin.coords.y, expected_ray.origin.coords.y, max_ulps = 2);
        assert_ulps_eq!(ray.origin.coords.z, expected_ray.origin.coords.z, max_ulps = 2);
```

Reply:
I'm pretty sure `assert_ulps_eq` *does* work with `Vector3`. The issue here is just that the literal values are not very precise. That's why I was using epsilon before.